charms/worker/k8s/src/token_distributor.py

```diff
@@ -275,24 +308,30 @@ def revoke_tokens(
             token_strategy (TokenStrategy): The strategy of token creation.
             token_type (ClusterTokenType, optional): The type of cluster token.
                 Defaults to ClusterTokenType.NONE.
-            to_remove (Set[ops.Unit], optional): Specific units to ensure its token is revoked
+            to_remove (ops.Unit, optional): unit to ensure its token is revoked
 
         Raises:
             ValueError: If an invalid token_strategy is provided.
         """
+        # any unit currently in the relation
+        all_units = relation.units
+        if self.charm.app == relation.app:
+            # include self in peer relations
+            all_units |= {self.charm.unit}
+
+        # any unit in the app_databag, which successfully recovered its token
         app_databag = relation.data[self.charm.app]
         joined = {self.charm.model.get_unit(str(u)) for u in app_databag}
-        if to_remove is not None:
-            remaining = joined - to_remove
-        else:
-            remaining = joined & (relation.units | {self.charm.unit})
-        remove = joined - remaining
+
+        # establish the remaining units
+        remove = {to_remove} if to_remove else (joined - all_units)
```

Comment:
I feel like you need to have a math minor to deal with the set theory in this change. 

Problem:  There's 2 sets of truths
1) The app_databag keeps up with a mapping of `units` -> `state + nodename`
2) the relation databag which tells us the current relation status
   * caveat:  on relation-departed hooks -- the unit that is departing is GONE!

Solutions:
1) if i'm told to remove one unit -- only try to remove that ONE unit -- ignore trying to reap all the others for now
2) If i'm not told which unit to remove, then we should only remove units in the `app_databag` that aren't in the current relation databag.  That's why it's **SO IMPORTANT** we put the `state + node` in the app_databag when we first make a token for someone.  If not, we might make a token, it gets used by a node to join, but it doesn't TELL us that it joined, so we can't remove it.